// grunt-we7-html2js/tasks/we7_html2js.js
/*
 * grunt-we7-html2js
 * https://github.com/donknap/grunt-we7-html2js
 *
 * Copyright (c) 2016 WeEngine Team
 * Licensed under the MIT license.
 */

'use strict';

module.exports = function(grunt) {

  var path = require('path');
  var minify = require('html-minifier').minify;

  var escapeContent = function(content, quoteChar, indentString) {
    var bsRegexp = new RegExp('\\\\', 'g');
    var quoteRegexp = new RegExp('\\' + quoteChar, 'g');
    var nlReplace = '\\n' + quoteChar + ' +\n' + indentString + indentString + quoteChar;
    return content.replace(bsRegexp, '\\\\').replace(quoteRegexp, '\\' + quoteChar).replace(/\r?\n/g, nlReplace);
  };

  // convert Windows file separator URL path separator
  var normalizePath = function(p) {
    if ( path.sep !== '/' ) {
      p = p.replace(/\\/g, '/');
    }
    return p;
  };

  // Warn on and remove invalid source files (if nonull was set).
  var existsFilter = function(filepath) {

    if (!grunt.file.exists(filepath)) {
      grunt.log.warn('Source file "' + filepath + '" not found.');
      return false;
    } else {
      return true;
    }
  };

  // return template content
  var getContent = function(filepath, quoteChar, indentString, htmlmin, process) {
    var content = grunt.file.read(filepath);

    // Process files as templates if requested.
    if (typeof process === "function") {
      content = process(content, filepath);
    } else if (process) {
      if (process === true) {
        process = {};
      }
      content = grunt.template.process(content, process);
    }

    if (Object.keys(htmlmin).length) {
      try {
        content = minify(content, htmlmin);
      } catch (err) {
        grunt.warn(filepath + '\n' + err);
      }
    }

    return escapeContent(content, quoteChar, indentString);
  };

  // compile a template to an angular module
  var compileTemplate = function(filepath, quoteChar, indentString, useStrict, htmlmin, process, indexRename) {
    var content = getContent(filepath, quoteChar, indentString, htmlmin, process);
    if (typeof indexRename === "function") {
      filepath = indexRename(filepath);
    }
    var module = 'templates[' + quoteChar + filepath + quoteChar + '] = ' + quoteChar + content + quoteChar + ';';

    return module;
  };

  // compile a template to an angular module
  var compileCoffeeTemplate = function(filepath, quoteChar, indentString, htmlmin, process, indexRename) {
    var content = getContent(filepath, quoteChar, indentString, htmlmin, process);
    if (typeof indexRename === "function") {
      filepath = indexRename(filepath);
    }
    var module = 'templates[' + quoteChar + filepath + quoteChar + '] = ' + quoteChar + content + quoteChar + ';';

    return module;
  };
  grunt.registerMultiTask('we7_html2js', 'Html to Javascript', function() {

    var options = this.options({
      quoteChar: '"',
      globalname: 'templates',
      fileHeaderString: '',
      fileFooterString: '',
      indentString: '  ',
      target: 'js',
      htmlmin: {},
      process: false,
      indexRename : false
    });

    var counter = 0;

    // generate a separate module
    this.files.forEach(function(f) {

      var modules = f.src.filter(existsFilter).map(function(filepath) {

        if (options.target === 'js') {
          return compileTemplate(filepath, options.quoteChar, options.indentString, options.useStrict, options.htmlmin, options.process, options.indexRename);
        } else if (options.target === 'coffee') {
          return compileCoffeeTemplate(filepath, options.quoteChar, options.indentString, options.htmlmin, options.process, options.indexRename);
        } else {
          grunt.fail.fatal('Unknow target "' + options.target + '" specified');
        }

      });

      counter += modules.length;
      modules  = modules.join('\n' + options.indentString);

      var fileHeader = options.fileHeaderString !== '' ? options.fileHeaderString + '\n' : '';
      var fileFooter = options.fileFooterString !== '' ? options.fileFooterString + '\n' : '';
      var strict = (options.useStrict) ? options.indentString + options.quoteChar + 'use strict' + options.quoteChar + ';\n' : '';
      var bundle = "";
      var globalTemplates = 'this.' + options.globalname;
      var globalExports = globalTemplates + ' = ' + globalTemplates + ' || {}';

      if (options.target === 'js') {
        bundle = ';(function (templates, undefined) {\n' + strict + options.indentString + modules + '\n})(' + globalExports + ');';
      } else if (options.target === 'coffee') {
        bundle = '((templates, _undefined) -> \n' + strict + options.indentString + modules + '\n' + options.indentString + 'return;\n)(' + globalExports + ');';
      }

      grunt.file.write(f.dest, grunt.util.normalizelf(fileHeader + bundle + fileFooter));
    });
    //Just have one output, so if we making thirty files it only does one line
    grunt.log.writeln("Successfully converted "+(""+counter).green +
                      " html templates to " + options.target + ".");
  });

};
